onnxruntime/core/session/model_builder_c_api.cc

```diff
@@ -0,0 +1,347 @@
+#include "core/session/model_builder_api.h"
+#include "core/session/ort_apis.h"
+#include "core/session/ort_env.h"
+#include "core/session/utils.h"
+
+using namespace onnxruntime;
+
+ORT_API_STATUS_IMPL(OrtModelBuilderAPI::CreateValueInfo, _In_ const char* name, _In_ const OrtTypeInfo* type_info,
+                    _Outptr_ OrtValueInfo** value_info) {
+  API_IMPL_BEGIN
+  if (name == nullptr || *name == '\0') {
+    return OrtApis::CreateStatus(ORT_INVALID_ARGUMENT, "name cannot be null or empty string");
+  }
+
+  if (type_info == nullptr) {
+    return OrtApis::CreateStatus(ORT_INVALID_ARGUMENT, "type_info cannot be null");
+  }
+
+  if (type_info->type != ONNX_TYPE_TENSOR) {
+    return OrtApis::CreateStatus(ORT_FAIL, "Only tensor types are supported currently");
+  }
+
+  if (type_info->tensor_type_info == nullptr) {
+    return OrtApis::CreateStatus(ORT_INVALID_ARGUMENT, "tensor_type_info cannot be null");
+  }
+
+  auto vi = std::make_unique<OrtValueInfo>();
+  vi->name = name;
+  vi->type_info = type_info->Clone();
+
+  *value_info = vi.release();
+
+  return nullptr;
+  API_IMPL_END
+}
+
+ORT_API_STATUS_IMPL(OrtModelBuilderAPI::GetValueInfoName, _In_ const OrtValueInfo* value_info, _Out_ const char** name) {
+  API_IMPL_BEGIN
+  *name = value_info->name.c_str();
+  return nullptr;
+  API_IMPL_END
+}
+ORT_API_STATUS_IMPL(OrtModelBuilderAPI::GetValueInfoTypeInfo, _In_ const OrtValueInfo* value_info, _Outptr_ const OrtTypeInfo** type_info) {
+  API_IMPL_BEGIN
+
+  *type_info = value_info->type_info.get();
+
+  return nullptr;
+  API_IMPL_END
+}
+
+ORT_API(void, OrtModelBuilderAPI::ReleaseValueInfo, _Frees_ptr_opt_ OrtValueInfo* value_info) {
+  delete value_info;
+}
+
+ORT_API_STATUS_IMPL(OrtModelBuilderAPI::CreateNode, const char* operator_name, const char* domain_name,
+                    _In_ const char* node_name,
+                    _In_reads_(input_names_len) const char* const* input_names, size_t input_names_len,
+                    _In_reads_(output_names_len) const char* const* output_names, size_t output_names_len,
+                    _In_reads_(attribs_len) _Inout_opt_ OrtOpAttr** attributes, _In_opt_ size_t attribs_len,
+                    _Outptr_ OrtNode** node) {
+  API_IMPL_BEGIN
+  auto n = std::make_unique<OrtNode>();
+  n->operator_name = operator_name;
+  n->domain_name = domain_name == kOnnxDomainAlias ? kOnnxDomain : domain_name;
+  n->node_name = node_name;
+
+  n->input_names.reserve(input_names_len);
+  for (size_t i = 0; i < input_names_len; ++i) {
+    n->input_names.push_back(input_names[i]);
+  }
+
+  n->output_names.reserve(output_names_len);
+  for (size_t i = 0; i < output_names_len; ++i) {
+    n->output_names.push_back(output_names[i]);
+  }
+
+  if (attributes != nullptr) {
+    n->attributes.reserve(attribs_len);
+    for (size_t i = 0; i < attribs_len; ++i) {
+      n->attributes.push_back(*reinterpret_cast<const ONNX_NAMESPACE::AttributeProto*>(attributes[i]));
```

Comment:
Should we call ReleaseOpAttr after it's copied into the node so the user doesn't have to? Would be more consistent with the rest of the API to 'take ownership' of them.

Reply:
The same question applies to the CreateXXXTypeInfo/CreateValueInfo calls. 

Those felt a little more re-usable (e.g. if you were constructing a model with KV cache you're going to be using the same TypeInfo for multiple inputs/outputs.) but maybe it's better overall to have a consistent pattern of ownership transferring when you add to a containing class instead of taking a copy in some places.

Reply:
> Those felt a little more re-usable (e.g. if you were constructing a model with KV cache you're going to be using the same TypeInfo for multiple inputs/outputs.)

`OpAttr` might also be re-used?

>  but maybe it's better overall to have a consistent pattern of ownership transferring when you add to a containing class instead of taking a copy in some places.

I feel the consistency is for those `AddXxxToXxx()` methods who do ownership transferring? `SetGraphInputs()`/`SetGrapOutputs()` also take ownership (I was not aware of it when I initially used the API), should they be renamed to `AddInputsToGraph()`/`AddOutputsToGraph()`?

Reply:
`OpAttr` _could_ be re-used but that seems less likely. e.g. are you going to keep an OpAttr value around and look it up to re-use in a later node that _might_ need the same value?

My concern with AddInputsToGraph would be that it doesn't accumulate unlike the others, so 'set' is a better description.

Reply:
> OpAttr could be re-used but that seems less likely. e.g. are you going to keep an OpAttr value around and look it up to re-use in a later node that might need the same value?

WebNN POC code doesn't keep an OpAttr.

> My concern with AddInputsToGraph would be that it doesn't accumulate unlike the others, so 'set' is a better description.

That's good point. I am fine with `SetGraphInputs()`/`SetGrapOutputs()`.